R/benchmark-dataframe.R

```diff
@@ -0,0 +1,113 @@
+#' A vector of benchmark attribute names run on `ursa-i9-9960x`
+#'
+#' @export
+URSA_I9_9960X_R_BENCHMARK_NAMES <- c(
+  "dataframe-to-table",  # `df_to_table`
+  "file-read",
+  "file-write",
+  "partitioned-dataset-filter",  # `dataset_taxi_parquet`
+  "wide-dataframe",  # not actually an R benchmark
+  "tpch"  # `tpc_h`
+)
+
+#' A classed dataframe of benchmarks for running
+#'
+#' @param benchmarks A list with elements of class `Benchmark`
+#' @param parameters Optional. A list of dataframes of parameter combinations to
+#' run as generated by [default_params()]. If null, defaults will be generated
+#' when [run()] is called.
+#'
+#' @return A classed dataframe with `name` (benchmark attribute, not object name),
+#' `benchmark`, and `params` columns
+#'
+#' @export
+BenchmarkDataFrame <- function(benchmarks, parameters) {
+  lapply(benchmarks, function(bm) stopifnot(
+    "All elements of `benchmarks` are not of class `Benchmark`!" = inherits(bm, "Benchmark")
+  ))
+
+  bm_names <- vapply(benchmarks, function(bm) bm$name, character(1))
+  if (missing(parameters)) {
+    parameters <- rep(list(NULL), length = length(benchmarks))
+  }
+
+  structure(
+    tibble::tibble(
+      name = bm_names,
+      benchmark = benchmarks,
+      parameters = parameters
+    ),
+    class = c("BenchmarkDataFrame", "tbl_df", "tbl", "data.frame")
+  )
+}
+
+
+#' @export
+format.BenchmarkDataFrame <- function(x, ...) {
+  c("# <BenchmarkDataFrame>", NextMethod())
+}
+
+
+#' Get a list of benchmarks in a package
+#'
+#' @param package String of package name in which to find benchmarks
+#'
+#' @return An instance of [BenchmarkDataFrame] with all the benchmarks contained
+#' by a package
+#'
+#' @export
+get_package_benchmarks <- function(package = "arrowbench") {
+  nms <- getNamespaceExports(package)
+  objs <- mget(nms, envir = getNamespace(package))
+  bms <- Filter(function(x) inherits(x, "Benchmark"), objs)
+  BenchmarkDataFrame(benchmarks = bms)
+}
```

Comment:
This is some introspective R dark magic, but because these lines can only run after this package is installed, I'm fairly confident they're safe, and they keep us from needing to maintain a list of available benchmarks, which I like a lot.

Reply:
this is cool! So is there no way to override the default_params when you use this function?

Reply:
Good question, and no, I went with making it explicit by defining `default_params.BenchmarkDataFrame()` so you can do 

```r
get_package_benchmarks() |>
    default_params(cpu_count = arrow::cpu_count()) |>  # or whatever
    run()
```

Or you can put kwargs directly in `run()` and let the params dataframe be generated on the fly, but from a workflow perspective it was definitely useful to be able to see the params before running anything.

The big caveat here is that this is only really useful for params relevant to all the benchmarks in the dataframe, i.e. things like `cpu_count`. So it becomes way more important that our defaults are right (which I did a lot of work in the last PR to correct at least for the ones we're running regularly). For anything else, you'll probably need to assign to a subset with

```r
bm_df <- get_package_benchmarks()
bm_df$parameters[bm_df$name == "read_json"] <- default_params(read_json, source = c("fanniemae_2016", "nyctaxi_2010"))
```
or whatever. I don't love this, but don't see a better alternative, really.

Reply:
This is really nice and nice that it works with `load_all()`. Sometimes that can be a little hairy.